anvil/src/eth/backend/mem/mod.rs

```diff
@@ -1086,36 +1097,79 @@ impl Backend {
+    pub async fn with_database_at<F, T>(
+        &self,
+        block_request: Option<BlockRequest>,
+        f: F,
+    ) -> Result<T, BlockchainError>
     where
-        F: FnOnce(Box<dyn MaybeHashDatabase + '_>) -> T,
+        F: FnOnce(Box<dyn MaybeHashDatabase + '_>, BlockEnv) -> T,
     {
+        let block_number = match block_request {
+            Some(BlockRequest::Pending(pool_transactions)) => {
+                let result = self
+                    .with_pending_block(pool_transactions, |state, block| {
+                        let block = block.block;
+                        let block = BlockEnv {
+                            number: block.header.number,
+                            coinbase: block.header.beneficiary,
+                            timestamp: block.header.timestamp.into(),
+                            difficulty: block.header.difficulty,
+                            basefee: block.header.base_fee_per_gas.unwrap_or_default(),
+                            gas_limit: block.header.gas_limit,
+                        };
+                        f(state, block)
+                    })
+                    .await;
+                return Ok(result)
+            }
+            Some(BlockRequest::Number(bn)) => Some(BlockNumber::Number(bn)),
+            None => None,
+        };
         let block_number: U256 = self.convert_block_number(block_number).into();
 
         if block_number < self.env.read().block.number {
             let states = self.states.read();
-            if let Some(state) =
-                self.hash_for_block_number(block_number.as_u64()).and_then(|hash| states.get(&hash))
+            return if let Some((state, block)) = self
+                .get_block(block_number.as_u64())
+                .and_then(|block| Some((states.get(&block.header.hash())?, block)))
             {
-                return f(Box::new(state))
+                let block = BlockEnv {
+                    number: block.header.number,
+                    coinbase: block.header.beneficiary,
+                    timestamp: block.header.timestamp.into(),
+                    difficulty: block.header.difficulty,
+                    basefee: block.header.base_fee_per_gas.unwrap_or_default(),
+                    gas_limit: block.header.gas_limit,
+                };
+                Ok(f(Box::new(state), block))
+            } else {
+                warn!(target: "backend", "Not historic state found for block={}", block_number);
+                Err(BlockchainError::BlockOutOfRange(
+                    self.env.read().block.number.as_u64(),
+                    block_number.as_u64(),
+                ))
             }
         }
         let db = self.db.read().await;
-        f(Box::new(&*db))
+        let block = self.env().read().block.clone();
+        // @todo Do we want the `Latest` block to have an increased timestamp?
+        // block.timestamp = self.time.current_call_timestamp().into();
```

Comment:
"Latest", refers to the last mined block, so not updating the timestamp here actually fixes a bug.

however this surfaces another bug we had in forking mode that should be fixed by: https://github.com/foundry-rs/foundry/pull/2711
which sets the timestamp for the env properly

suggesting rm these lines
```suggestion
```